### PR TITLE
TVIST1-725: Allow deletion of hearing posts

### PR DESCRIPTION
**Pull request by jekuaitk**
https://jira.itkdev.dk/browse/TVIST1-725

* Allows deletion of created hearing post requests and responses
* Sends hearing post response receipt upon approval rather than creation.

### CODE REVIEW

File: src/Controller/HearingController.php [hunk 1]
```diff
@@ -369,13 +370,45 @@ public function hearingPostRequestEdit(CaseEntity $case, DocumentRepository $doc
...
+        if ($hearingPost->getSendReceipt()) {
+            $sender = $hearingPost->getSender();
+            $digitalPostRecipients[] = (new DigitalPost\Recipient())
+                ->setName($sender->getName())
+                ->setIdentifierType($sender->getIdentification()->getType())
+                ->setIdentifier($sender->getIdentification()->getIdentifier())
+                ->setAddress($sender->getAddress())
+            ;
+            $case = $hearingPost->getHearing()?->getCaseEntity();
+            $template = $case?->getBoard()?->getReceiptHearingPost();
```
Comment (rimi-itk) [inline src/Controller/HearingController.php:391]: `$case` and `$template` may be `null` so you should check for this before using the values in the code below.

File: src/Logging/EntityListener/CaseDocumentRelationListener.php [hunk 2]
```diff
@@ -17,6 +17,11 @@ public function postUpdate(CaseDocumentRelation $relation, LifecycleEventArgs $a
         $this->logActivity('Document updated', $args);
     }
 
+    public function preRemove(CaseDocumentRelation $relation, LifecycleEventArgs $args)
+    {
+        $this->logActivity('Document deleted', $args);
+    }
+
```
Comment (rimi-itk) [inline src/Logging/EntityListener/CaseDocumentRelationListener.php:24]: 👍

File: src/Logging/EntityListener/HearingPostListener.php [hunk 3]
```diff
@@ -15,12 +18,32 @@ public function __construct(Security $security)
...
+    private function getHearingPostType(HearingPost $hearingPost)
+    {
+        if ($hearingPost instanceof HearingPostRequest) {
+            return 'Hearing post request';
+        } elseif ($hearingPost instanceof HearingPostResponse) {
+            return 'Hearing post response';
+        }
+
+        $message = sprintf('Unhandled hearing post type %s found.', get_class($hearingPost));
+        throw new ItkDevLoggingException($message);
```
Comment (rimi-itk) [inline src/Logging/EntityListener/HearingPostListener.php:46]: Why don't you use your cool `return match(true) { … }` trick here?

File: templates/case/hearing/index.html.twig [hunk 4]
```diff
@@ -169,11 +169,21 @@
                                                 <form action="{{ path('case_hearing_post_forward', {case: case.id, hearingPost: post.id}) }}" method="post" style="display: inline-block">
                                                     <button class="btn btn-warning">{% trans %}Send{% endtrans %}</button>
                                                 </form>
+                                                <form class="btn" style="margin:0; padding:0; display:inline;" method="post" action="{{ path('case_hearing_post_delete', {case: case.id, hearingPost: post.id}) }}" onsubmit="return confirm('{% trans %}Are you sure you want to delete this item?{% endtrans %}');">
```
Comment (rimi-itk) [inline templates/case/hearing/index.html.twig:172]: The inline styling should be moved to a css file.

File: templates/case/hearing/index.html.twig [hunk 5]
```diff
@@ -169,11 +169,21 @@
...
+                                                    <input type="hidden" name="_token" value="{{ csrf_token('delete' ~ post.id) }}">
+                                                    <button class="btn btn-danger">{% trans %}Delete{% endtrans %}</button>
+                                                </form>
                                             {% endif %}
                                             {% if post.approvedOn is defined and not post.getApprovedOn %}
                                                 <form action="{{ path('case_hearing_post_approve', {case: case.id, hearingPost: post.id}) }}" method="post" style="display: inline-block">
                                                     <button class="btn btn-warning">{% trans %}Approve{% endtrans %}</button>
                                                 </form>
+                                                <form class="btn" style="margin:0; padding:0; display:inline;" method="post" action="{{ path('case_hearing_post_delete', {case: case.id, hearingPost: post.id}) }}" onsubmit="return confirm('{% trans %}Are you sure you want to delete this item?{% endtrans %}');">
```
Comment (rimi-itk) [inline templates/case/hearing/index.html.twig:182]: The duplicated inline styling should be moved to a css file.

File: src/Controller/HearingController.php [hunk 6]
```diff
@@ -462,4 +495,48 @@ public function resumeHearing(CaseEntity $case, Hearing $hearing): Response
...
+            // Remove hearing post and document
+            $this->entityManager->remove($hearingPost);
+            $this->entityManager->remove($document);
+            $this->entityManager->flush();
+
+            if ($hearingPost instanceof HearingPostResponse) {
+                $this->addFlash('success', new TranslatableMessage('Hearing post response deleted', [], 'case'));
+            } elseif ($hearingPost instanceof HearingPostRequest) {
+                $this->addFlash('success', new TranslatableMessage('Hearing post request deleted', [], 'case'));
+            }
```
Comment (rimi-itk) [inline src/Controller/HearingController.php:537]: See https://github.com/itk-dev/naevnssekretariatet/blob/develop/src/Controller/AgendaItemController.php#L68-L72 for another (better?) way to do this.

File: src/Controller/HearingController.php [hunk 7]
```diff
@@ -462,4 +495,48 @@ public function resumeHearing(CaseEntity $case, Hearing $hearing): Response
...
+
+            if (null !== $relation) {
+                $this->entityManager->remove($relation);
+                $this->entityManager->flush();
+            }
+
+            // Remove file
+            $documentUploader->deleteDocumentFile($document);
+
+            // Set foreign keys between hearing post and document to null
```
Comment (rimi-itk) [inline src/Controller/HearingController.php:520]: Consider rewording to “Remove references between post and document”, say. “Foreign keys” is something that the ORM handles and we don't really care about the techical details in our code.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: src/Controller/HearingController.php [hunk 8]
```diff
@@ -462,4 +495,48 @@ public function resumeHearing(CaseEntity $case, Hearing $hearing): Response
...
+            $hearingPost->getDocument()->setHearingPost(null);
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply src/Controller/HearingController.php:521]: Fixed.

File: src/Controller/HearingController.php (hunk 1 above)
Comment (jekuaitk) [reply src/Controller/HearingController.php:391]: Added check.

File: src/Logging/EntityListener/HearingPostListener.php (hunk 3 above)
Comment (jekuaitk) [reply src/Logging/EntityListener/HearingPostListener.php:46]: Used cool trick 😎

File: templates/case/hearing/index.html.twig (hunk 4 above)
Comment (jekuaitk) [reply templates/case/hearing/index.html.twig:172]: Moved to css file.

File: templates/case/hearing/index.html.twig (hunk 5 above)
Comment (jekuaitk) [reply templates/case/hearing/index.html.twig:182]: Moved to css file.

File: src/Controller/HearingController.php (hunk 7 above)
Comment (jekuaitk) [reply src/Controller/HearingController.php:520]: Reworded.

File: src/Controller/HearingController.php (hunk 6 above)
Comment (jekuaitk) [reply src/Controller/HearingController.php:537]: Refactored to use proposed method.

File: templates/case/hearing/index.html.twig [hunk 9]
```diff
@@ -169,7 +169,7 @@
                                                 <form action="{{ path('case_hearing_post_forward', {case: case.id, hearingPost: post.id}) }}" method="post" style="display: inline-block">
                                                     <button class="btn btn-warning">{% trans %}Send{% endtrans %}</button>
                                                 </form>
-                                                <form class="btn" style="margin:0; padding:0; display:inline;" method="post" action="{{ path('case_hearing_post_delete', {case: case.id, hearingPost: post.id}) }}" onsubmit="return confirm('{% trans %}Are you sure you want to delete this item?{% endtrans %}');">
+                                                <form class="btn inline-delete-form" method="post" action="{{ path('case_hearing_post_delete', {case: case.id, hearingPost: post.id}) }}" onsubmit="return confirm('{% trans %}Are you sure you want to delete this item?{% endtrans %}');">
```
Comment (rimi-itk) [inline templates/case/hearing/index.html.twig:172]: Why does the form have the `btn` class?

**Review by rimi-itk — APPROVED** [review]

File: templates/case/hearing/index.html.twig (hunk 9 above)
Comment (jekuaitk) [reply templates/case/hearing/index.html.twig]: Removed.

**Pull request merged**